src/server/tablet_snap.rs

```diff
@@ -1,82 +1,157 @@
+//!
+//! sender                            receiver
+//! send snapshot meta    ---->     receive snapshot meta
+//! extra snapshot preview          collect cache meta
+//! send all preview      ---->     receive preview and clean up miss cache
+//! files receive files list      <-----  send missing file list
+//! send missing files    ---->     receive missing files
+//! close sender          ---->     persist snapshot and report to raftstore
+//! wait for receiver       <-----  close sender
+//! finish
+
 #[cfg(any(test, feature = "testexport"))]
 use std::io;
 use std::{
-    convert::{TryFrom, TryInto},
+    cmp,
+    convert::TryFrom,
+    fmt::Debug,
     fs::{self, File},
-    io::{Read, Write},
+    io::{BorrowedBuf, Read, Seek, SeekFrom, Write},
+    path::Path,
     sync::{
         atomic::{AtomicUsize, Ordering},
         Arc,
     },
     time::Duration,
 };
 
-use file_system::{IoType, WithIoType};
+use collections::HashMap;
+use engine_traits::{Checkpointer, KvEngine, TabletRegistry};
+use file_system::{IoType, OpenOptions, WithIoType};
 use futures::{
-    future::{Future, TryFutureExt},
+    future::{Future, FutureExt},
     sink::{Sink, SinkExt},
     stream::{Stream, StreamExt, TryStreamExt},
 };
 use grpcio::{
-    self, ChannelBuilder, ClientStreamingSink, Environment, RequestStream, RpcStatus,
-    RpcStatusCode, WriteFlags,
+    self, ChannelBuilder, DuplexSink, Environment, RequestStream, RpcStatus, RpcStatusCode,
+    WriteFlags,
 };
 use kvproto::{
-    raft_serverpb::{Done, RaftMessage, RaftSnapshotData, SnapshotChunk},
+    raft_serverpb::{
+        RaftMessage, RaftSnapshotData, TabletSnapshotFileChunk, TabletSnapshotFileMeta,
+        TabletSnapshotPreview, TabletSnapshotRequest, TabletSnapshotResponse,
+    },
     tikvpb::TikvClient,
 };
 use protobuf::Message;
-use raftstore::store::snap::{TabletSnapKey, TabletSnapManager};
+use raftstore::store::snap::{ReceivingGuard, TabletSnapKey, TabletSnapManager};
 use security::SecurityManager;
 use tikv_kv::RaftExtension;
 use tikv_util::{
-    config::{Tracker, VersionTrack},
+    config::{ReadableSize, Tracker, VersionTrack},
     time::Instant,
     worker::Runnable,
 };
 use tokio::runtime::{Builder as RuntimeBuilder, Runtime};
 
 use super::{
     metrics::*,
-    snap::{Task, DEFAULT_POOL_SIZE, SNAP_CHUNK_LEN},
+    snap::{Task, DEFAULT_POOL_SIZE},
     Config, Error, Result,
 };
 use crate::tikv_util::{sys::thread::ThreadBuildWrapper, time::Limiter};
 
-struct RecvTabletSnapContext {
+const PREVIEW_CHUNK_LEN: usize = ReadableSize::kb(1).0 as usize;
+const PREVIEW_BATCH_SIZE: usize = 256;
+const FILE_CHUNK_LEN: usize = ReadableSize::mb(1).0 as usize;
+const CACHE_LINE: u64 = ReadableSize::mb(4).0;
+
+fn is_sst(file_name: &str) -> bool {
+    file_name.ends_with(".sst")
+}
+
+async fn read_to(f: &mut File, to: &mut Vec<u8>, size: usize, limiter: &Limiter) -> Result<()> {
+    // It's likely in page cache already.
+    limiter.consume(size / 2).await;
```

Comment:
The limiter is configured by `snap_max_write_bytes_per_sec` though. Read bandwidth should be implicitly limited by receiver end.

Reply:
For example, if the node is sending snapshot to 8 different nodes, then the read bandwidth can be as high as `snap_max_write_bytes_per_sec` * 8.

Reply:
How about a new config. It could be difficult to tune one without affecting another.